### PR TITLE
feature: add code for calibration

### PR DESCRIPTION
**Pull request by alxlyj**
## Project Robyn

As titled. We are adding the necessary interfaces and implementation for calibration component.
We also integrated the new addition from calibration to modeling.
These are also integrated and tested on a tutorial notebook. 

## Test Plan

Code flow sequence diagram on R side:

```mermaid
flowchart TD
    A[Start robyn_calibrate] --> B{Check if calibration_input exists<br/>and study is within window}
    B -->|No calibration_input or<br/>outside window| C[Warning: Running without calibration]
    B -->|Valid input and<br/>within window| D[Initialize calibration_input<br/>with NA columns]
    
    D --> E[Process each study in split_channels]
    
    subgraph study_loop[Study Processing Loop]
        E --> F[Get channels and parameters<br/>for current study]
        F --> G[Determine study position<br/>and calibration dates]
        G --> H[Process each channel]
        
        subgraph channel_loop[Channel Processing Loop]
            H --> I{Check calibration scope}
            
            I -->|immediate| J[Process immediate scope]
            J --> K1[Apply Adstock transformation]
            K1 --> L1{Check adstock type}
            L1 -->|geometric| M1[Apply geometric<br/>transformation]
            L1 -->|weibull| M2[Apply weibull<br/>transformation]
            M1 --> N1[Calculate immediate<br/>and total effects]
            M2 --> N1
            N1 --> O1[Apply Saturation<br/>Hill transformation]
            O1 --> P1[Calculate decomposition<br/>for immediate scope]
            
            I -->|total| J2[Process total scope]
            J2 --> P2[Get direct decomposition<br/>from xDecompVec]
        end
        
        P1 --> Q[Combine multiple<br/>channel results]
        P2 --> Q
        Q --> R[Update calibration_input<br/>with predictions]
    end
    
    R --> S[Calculate final metrics]
    S --> T[Generate liftCollect with<br/>scaled decomposition values]
    T --> U[Return liftCollect]

    subgraph calculations[Key Calculations]
        V1[mape_lift = decompAbsScaled - liftAbs / liftAbs]
        V2[calibrated_pct = decompAbsScaled / decompAbsTotalScaled]
        V3[decompAbsScaled = pred / decompDays * liftDays]
    end

    style study_loop fill:#f0f0f0,stroke:#333,stroke-width:2px
    style channel_loop fill:#e6e6e6,stroke:#666,stroke-width:2px
    style calculations fill:#e6f3ff,stroke:#0066cc,stroke-width:2px
```

Tutorial notebook results with calibration:

#### Calibration input:
image

#### Calibration Results:

image

Accuracy of calibration will be vetted in a future PR. 

image

### CODE REVIEW

**Comment by alxlyj** [comment]
for reference, this is what we are seeing for calibration input from the R side:

image

File: python/src/robyn/calibration/media_effect_calibration.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+                calibration_scores[channel] = score
+
+            except Exception as e:
+                print(f"Error calculating calibration for channel {channel}: {str(e)}")
+                # You might want to handle this differently based on your requirements
+                calibration_scores[channel] = float("inf")
+
+        return CalibrationResult(
+            channel_scores=calibration_scores,
+            calibration_constraint=0.05,  # Default constraint
```
Comment (sumane81) [inline python/src/robyn/calibration/media_effect_calibration.py:204]: This seems to be hard coded. CalibrationResult definition would be right place to have default value.

File: python/src/robyn/calibration/media_effect_calibration.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        self.media_transformation = MediaTransformation(hyperparameters)
+
+        # Convert date column to datetime if it's not already
+        date_col = self.mmm_data.mmmdata_spec.date_var
+        if not pd.api.types.is_datetime64_any_dtype(self.mmm_data.data[date_col]):
+            self.mmm_data.data[date_col] = pd.to_datetime(self.mmm_data.data[date_col])
+
+        self._validate_inputs()
+
+    def _split_channel(self, channel: str) -> List[str]:
```
Comment (sumane81) [inline python/src/robyn/calibration/media_effect_calibration.py:69]: calibration_input should contain individual channels. See if we can support combination of channel in calibration input itself.

For example - currently calibration_input is Dict[channel:, channelCalibrationData]. If we can have Dict[List[channel], channelCalibrationData].

File: python/src/robyn/calibration/media_effect_calibration.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+                    )
+                else:
+                    score = self._calculate_total_effect_score(
+                        total_pred, data.lift_abs, data.spend, channel_for_params
+                    )
+
+                calibration_scores[channel] = score
+
+            except Exception as e:
+                print(f"Error calculating calibration for channel {channel}: {str(e)}")
```
[the remark this answers is not in the document]
Comment (sumane81) [reply python/src/robyn/calibration/media_effect_calibration.py:198]: Overall need more logging of Info, debug and errors.

File: python/src/robyn/modeling/ridge_model_builder.py [hunk 4]
```diff
@@ -582,80 +531,90 @@ def _evaluate_model(
...
+            if lift_collect is not None:
+                lift_calibration = lift_collect
+                mape = float(lift_collect.get_mean_mape())
+                calibration_constraint = 0.1
+                if mape > calibration_constraint:
+                    decomp_rssd *= 1 + (mape - calibration_constraint)
+            else:
+                mape = 0.0
+        else:
+            mape = 0.0
```
Comment (sumane81) [inline python/src/robyn/modeling/ridge_model_builder.py:602]: nit; just add mape = 0.0 before if block so you don't need two else blocks above.

**Review by sumane81 — CHANGES_REQUESTED** [review]
Mostly minor comments.

File: python/src/robyn/calibration/media_effect_calibration.py (hunk 2 above)
Comment (alxlyj) [reply python/src/robyn/calibration/media_effect_calibration.py:75]: something like this?

calibration_data = {
    ["facebook"]: ChannelCalibrationData(  # Single channel as a list
        lift_start_date=pd.Timestamp("2023-01-01"),
        lift_end_date=pd.Timestamp("2023-01-31"),
        lift_abs=1000,
        spend=5000
    ),
    ["search", "tv"]: ChannelCalibrationData(  # Multiple channels as a list
        lift_start_date=pd.Timestamp("2023-02-01"),
        lift_end_date=pd.Timestamp("2023-02-28"),
        lift_abs=2000,
        spend=8000
    )
}
calibration_input = CalibrationInput(channel_data=calibration_data)

**Comment by alxlyj** [comment]
now supports combination of channel in calibration input itself.
For example - currently calibration_input is Dict[channel:, channelCalibrationData]. If we can have Dict[List[channel], channelCalibrationData].

i.e.:
calibration_data = {
["facebook"]: ChannelCalibrationData( # Single channel as a list
lift_start_date=pd.Timestamp("2023-01-01"),
lift_end_date=pd.Timestamp("2023-01-31"),
lift_abs=1000,
spend=5000
),
["search", "tv"]: ChannelCalibrationData( # Multiple channels as a list
lift_start_date=pd.Timestamp("2023-02-01"),
lift_end_date=pd.Timestamp("2023-02-28"),
lift_abs=2000,
spend=8000
)
}
calibration_input = CalibrationInput(channel_data=calibration_data)

File: python/src/robyn/calibration/media_effect_calibration.py (hunk 2 above)
Comment (sumane81) [reply python/src/robyn/calibration/media_effect_calibration.py:75]: Exactly.

File: python/src/robyn/data/validation/calibration_input_validation.py [hunk 5]
```diff
@@ -91,101 +41,237 @@ def check_obj_weights(self, objective_weights: List[float], refresh: bool) -> Va
...
         return ValidationResult(
-            status=len(error_details) == 0,
-            error_details=error_details,
-            error_message="\n".join(error_messages)
+            status=len(error_details) == 0, error_details=error_details, error_message="\n".join(error_messages)
         )
 
-    def validate(self) -> ValidationResult:
+    @staticmethod
+    def create_modified_calibration_input(
```
Comment (alxlyj) [inline python/src/robyn/data/validation/calibration_input_validation.py:210]: - util function to help test data modification and to help simplify test cases. 
- handle special test cases like non-existent channels

File: python/src/robyn/data/entities/calibration_input.py [hunk 6]
```diff
@@ -47,13 +41,26 @@ class CalibrationInput:
...
 
-    channel_data: Dict[str, ChannelCalibrationData] = field(default_factory=dict)
+    channel_data: Dict[Union[str, Tuple[str, ...]], ChannelCalibrationData] = field(default_factory=dict)
+
+    def __post_init__(self):
+        # Convert string keys with '+' to tuples if needed
+        new_channel_data = {}
+        for key, value in self.channel_data.items():
+            if isinstance(key, str) and "+" in key:
+                new_key = tuple(key.split("+"))
```
Comment (sumane81) [inline python/src/robyn/data/entities/calibration_input.py:55]: we shouldn't support option of combination of channel with concatenated string with "+" at all.

**Review by sumane81 — APPROVED** [review]
Looks good to me. Make sure to create a task for unit tests and accuracy confirmation.

**Pull request merged**